pkg/controller/framework/shared_informer.go

```diff
@@ -0,0 +1,280 @@
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/runtime"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+)
+
+// if you use this, there is one behavior change.  When you receive a notification, the cache will be AT LEAST as
+// current as the notification, but it MAY be more current.
+// You should NOT depend on the contents of the cache exactly matching the notification you've received in handler functions.
+// If there was a create, followed by a delete, the cache
+// would NOT have your item.  All controllers should already be deduping on their side (if you forget to brush your teeth
+// at bedtime, you don't do it twice in the morning), but I doubt many people have considered the problem
+// This has advantages over the broadcaster since it allows us to share a common cache across many controllers.
+// Extending the broadcaster would have required us keep duplicate caches for each watch.
+type SharedInformer interface {
+	// events to a single handler are delivered sequentially, but there is no coordinatino between different handlers
+	// You may NOT add a handler *after* the SharedInformer is running.  That will result in an error being returned.
+	AddEventHandler(handler ResourceEventHandler) error
+	GetStore() cache.Store
+	// GetController gives back a synthetic interface that "votes" to start the informer
+	GetController() ControllerInterface
+	Run(stopCh <-chan struct{})
+	HasSynced() bool
+}
+
+type SharedIndexInformer interface {
+	SharedInformer
+
+	AddIndexer(indexer cache.Indexer) error
+	GetIndexer() cache.Indexer
+}
+
+// NewSharedInformer creates a new instance for the listwatcher.
+// TODO: create a cache/factory of these at a higher level for the list all, watch all of a given resource that can
+// be shared amongst all consumers.
+func NewSharedInformer(lw cache.ListerWatcher, objType runtime.Object, resyncPeriod time.Duration) SharedInformer {
+	sharedInformer := &ShareableInformer{
+		processor: &shareableProcessor{},
+		store:     cache.NewStore(DeletionHandlingMetaNamespaceKeyFunc),
+	}
+
+	// This will hold incoming changes. Note how we pass store in as a
+	// KeyLister, that way resync operations will result in the correct set
+	// of update/delete deltas.
+	fifo := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, sharedInformer.store)
+
+	cfg := &Config{
+		Queue:            fifo,
+		ListerWatcher:    lw,
+		ObjectType:       objType,
+		FullResyncPeriod: resyncPeriod,
+		RetryOnError:     false,
+
+		Process: sharedInformer.HandleDeltas,
+	}
+	sharedInformer.controller = New(cfg)
+
+	return sharedInformer
+}
+
+type ShareableInformer struct {
+	store      cache.Store
+	controller *Controller
+
+	processor *shareableProcessor
+
+	runVotes    sync.WaitGroup
+	started     bool
+	startedLock sync.Mutex
+}
+
+type votingController struct {
+	informer *ShareableInformer
+}
+
+func (v *votingController) Run(stopCh <-chan struct{}) {
+	v.informer.runVotes.Done()
```

Comment:
>
I'd prefer an interface that works for both shared and unshared things.

Yes - this one I definitely agree with. That's why I initially wanted to move Done() to a different place, not renaming this one. It's definitely a good goal to have.

>
Run() is always a vote,

That's something I completely disagree with. For me Run() is run it, not I vote for running, but who knows when and if it will be running.
[also adding @lavalamp for thoughts about it]